// lisk-framework/dist-node/modules/fee/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFeeTokenIDResponseSchema = exports.getMinFeePerByteResponseSchema = exports.configSchema = void 0;
exports.configSchema = {
    $id: '/fee/config',
    type: 'object',
    properties: {
        feeTokenID: {
            type: 'string',
            format: 'hex',
        },
        minFeePerByte: {
            type: 'integer',
            format: 'uint32',
        },
        maxBlockHeightZeroFeePerByte: {
            type: 'integer',
            format: 'uint32',
        },
        feePoolAddress: {
            type: 'string',
            format: 'lisk32',
        },
    },
    required: ['feeTokenID', 'minFeePerByte'],
};
exports.getMinFeePerByteResponseSchema = {
    $id: '/fee/endpoint/getMinFeePerByteResponse',
    type: 'object',
    properties: {
        minFeePerByte: {
            type: 'integer',
            format: 'uint32',
        },
    },
    required: ['minFeePerByte'],
};
exports.getFeeTokenIDResponseSchema = {
    $id: '/fee/endpoint/getFeeTokenIDResponseSchema',
    type: 'object',
    properties: {
        feeTokenID: {
            type: 'string',
            format: 'hex',
        },
    },
    required: ['feeTokenID'],
};
