// lisk-framework/dist-node/modules/nft/events/all_nfts_from_chain_support_removed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllNFTsFromChainSupportRemovedEvent = exports.allNFTsFromChainSupportRemovedEventSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.allNFTsFromChainSupportRemovedEventSchema = {
    $id: '/nft/events/allNFTsFromChainSupportRemoved',
    type: 'object',
    required: ['chainID'],
    properties: {
        chainID: {
            dataType: 'bytes',
            minLength: constants_1.LENGTH_CHAIN_ID,
            maxLength: constants_1.LENGTH_CHAIN_ID,
            fieldNumber: 1,
        },
    },
};
class AllNFTsFromChainSupportRemovedEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.allNFTsFromChainSupportRemovedEventSchema;
    }
    log(ctx, chainID) {
        this.add(ctx, { chainID }, [chainID]);
    }
}
exports.AllNFTsFromChainSupportRemovedEvent = AllNFTsFromChainSupportRemovedEvent;
