// lisk-framework/dist-node/modules/nft/events/ccm_transfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CcmTransferEvent = exports.ccmTransferEventSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.ccmTransferEventSchema = {
    $id: '/nft/events/ccmTransfer',
    type: 'object',
    required: [
        'senderAddress',
        'recipientAddress',
        'nftID',
        'receivingChainID',
        'sendingChainID',
        'result',
    ],
    properties: {
        senderAddress: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 1,
        },
        recipientAddress: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 2,
        },
        nftID: {
            dataType: 'bytes',
            minLength: constants_1.LENGTH_NFT_ID,
            maxLength: constants_1.LENGTH_NFT_ID,
            fieldNumber: 3,
        },
        receivingChainID: {
            dataType: 'bytes',
            minLength: constants_1.LENGTH_CHAIN_ID,
            maxLength: constants_1.LENGTH_CHAIN_ID,
            fieldNumber: 4,
        },
        sendingChainID: {
            dataType: 'bytes',
            minLength: constants_1.LENGTH_CHAIN_ID,
            maxLength: constants_1.LENGTH_CHAIN_ID,
            fieldNumber: 5,
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 6,
        },
    },
};
class CcmTransferEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.ccmTransferEventSchema;
    }
    log(ctx, data) {
        this.add(ctx, { ...data, result: 0 }, [
            data.senderAddress,
            data.recipientAddress,
        ]);
    }
    error(ctx, data, result) {
        this.add(ctx, { ...data, result }, [data.senderAddress, data.recipientAddress], true);
    }
}
exports.CcmTransferEvent = CcmTransferEvent;
