// lisk-framework/dist-node/modules/nft/events/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateEvent = exports.createEventSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.createEventSchema = {
    $id: '/nft/events/create',
    type: 'object',
    required: ['address', 'nftID', 'result'],
    properties: {
        address: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 1,
        },
        nftID: {
            dataType: 'bytes',
            minLength: constants_1.LENGTH_NFT_ID,
            maxLength: constants_1.LENGTH_NFT_ID,
            fieldNumber: 2,
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 3,
        },
    },
};
class CreateEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.createEventSchema;
    }
    log(ctx, data) {
        this.add(ctx, { ...data, result: 0 }, [
            data.address,
            data.nftID,
        ]);
    }
}
exports.CreateEvent = CreateEvent;
