// lisk-framework/dist-node/modules/nft/events/lock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LockEvent = exports.lockEventSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.lockEventSchema = {
    $id: '/nft/events/lock',
    type: 'object',
    required: ['module', 'nftID', 'result'],
    properties: {
        module: {
            dataType: 'string',
            minLength: constants_1.MIN_LENGTH_MODULE_NAME,
            maxLength: constants_1.MAX_LENGTH_MODULE_NAME,
            fieldNumber: 1,
        },
        nftID: {
            dataType: 'bytes',
            minLength: constants_1.LENGTH_NFT_ID,
            maxLength: constants_1.LENGTH_NFT_ID,
            fieldNumber: 2,
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 3,
        },
    },
};
class LockEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.lockEventSchema;
    }
    log(ctx, data) {
        this.add(ctx, { ...data, result: 0 }, [
            Buffer.from(data.module),
            data.nftID,
        ]);
    }
    error(ctx, data, result) {
        this.add(ctx, { ...data, result }, [Buffer.from(data.module), data.nftID], true);
    }
}
exports.LockEvent = LockEvent;
