// lisk-framework/dist-node/modules/nft/events/recover.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecoverEvent = exports.recoverEventSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.recoverEventSchema = {
    $id: '/nft/events/recover',
    type: 'object',
    required: ['terminatedChainID', 'nftID', 'result'],
    properties: {
        terminatedChainID: {
            dataType: 'bytes',
            minLength: constants_1.LENGTH_CHAIN_ID,
            maxLength: constants_1.LENGTH_CHAIN_ID,
            fieldNumber: 1,
        },
        nftID: {
            dataType: 'bytes',
            minLength: constants_1.LENGTH_NFT_ID,
            maxLength: constants_1.LENGTH_NFT_ID,
            fieldNumber: 2,
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 3,
        },
    },
};
class RecoverEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.recoverEventSchema;
    }
    log(ctx, data) {
        this.add(ctx, { ...data, result: 0 }, [data.nftID]);
    }
    error(ctx, data, result) {
        this.add(ctx, { ...data, result }, [data.nftID], true);
    }
}
exports.RecoverEvent = RecoverEvent;
