// lisk-framework/dist-node/modules/nft/events/set_attributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SetAttributesEvent = exports.setAttributesEventSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.setAttributesEventSchema = {
    $id: '/nft/events/setAttributes',
    type: 'object',
    required: ['nftID', 'attributes', 'result'],
    properties: {
        nftID: {
            dataType: 'bytes',
            minLength: constants_1.LENGTH_NFT_ID,
            maxLength: constants_1.LENGTH_NFT_ID,
            fieldNumber: 1,
        },
        attributes: {
            dataType: 'bytes',
            fieldNumber: 2,
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 3,
        },
    },
};
class SetAttributesEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.setAttributesEventSchema;
    }
    log(ctx, data) {
        this.add(ctx, { ...data, result: 0 }, [data.nftID]);
    }
    error(ctx, data, result) {
        this.add(ctx, { ...data, result }, [data.nftID], true);
    }
}
exports.SetAttributesEvent = SetAttributesEvent;
