// lisk-framework/dist-node/modules/token/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EMPTY_BYTES = exports.defaultConfig = exports.ESCROW_ACCOUNT_INITIALIZATION_FEE = exports.USER_ACCOUNT_INITIALIZATION_FEE = exports.LOCAL_ID_LSK = exports.TOKEN_ID_LENGTH = exports.LOCAL_ID_LENGTH = exports.HASH_LENGTH = exports.CHAIN_ID_LENGTH = exports.MIN_RETURN_FEE_PER_BYTE_BEDDOWS = exports.CCM_STATUS_PROTOCOL_VIOLATION = exports.CCM_STATUS_TOKEN_NOT_SUPPORTED = exports.CCM_STATUS_OK = exports.MAX_TRANSACTION_AMOUNT = exports.MAX_MODULE_NAME_LENGTH = exports.MIN_MODULE_NAME_LENGTH = exports.MAX_DATA_LENGTH = exports.ADDRESS_LENGTH = exports.CROSS_CHAIN_COMMAND_NAME_TRANSFER = void 0;
exports.CROSS_CHAIN_COMMAND_NAME_TRANSFER = 'transferCrossChain';
exports.ADDRESS_LENGTH = 20;
exports.MAX_DATA_LENGTH = 64;
exports.MIN_MODULE_NAME_LENGTH = 1;
exports.MAX_MODULE_NAME_LENGTH = 32;
exports.MAX_TRANSACTION_AMOUNT = '9223372036854775807';
exports.CCM_STATUS_OK = 0;
exports.CCM_STATUS_TOKEN_NOT_SUPPORTED = 64;
exports.CCM_STATUS_PROTOCOL_VIOLATION = 65;
exports.MIN_RETURN_FEE_PER_BYTE_BEDDOWS = BigInt(1000);
exports.CHAIN_ID_LENGTH = 4;
exports.HASH_LENGTH = 32;
exports.LOCAL_ID_LENGTH = 4;
exports.TOKEN_ID_LENGTH = exports.CHAIN_ID_LENGTH + exports.LOCAL_ID_LENGTH;
exports.LOCAL_ID_LSK = Buffer.alloc(exports.LOCAL_ID_LENGTH, 0);
exports.USER_ACCOUNT_INITIALIZATION_FEE = BigInt(5000000);
exports.ESCROW_ACCOUNT_INITIALIZATION_FEE = BigInt(5000000);
exports.defaultConfig = {
    userAccountInitializationFee: exports.USER_ACCOUNT_INITIALIZATION_FEE.toString(),
    escrowAccountInitializationFee: exports.ESCROW_ACCOUNT_INITIALIZATION_FEE.toString(),
};
exports.EMPTY_BYTES = Buffer.alloc(0);
