// lisk-framework/dist-node/modules/nft/commands/transfer_cross_chain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferCrossChainCommand = void 0;
const schemas_1 = require("../schemas");
const base_command_1 = require("../../base_command");
const state_machine_1 = require("../../../state_machine");
class TransferCrossChainCommand extends base_command_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.schema = schemas_1.crossChainTransferParamsSchema;
    }
    init(args) {
        this._internalMethod = args.internalMethod;
    }
    async verify(context) {
        const { params } = context;
        const { senderAddress } = context.transaction;
        try {
            await this._internalMethod.verifyTransferCrossChain(context.getMethodContext(), senderAddress, params.nftID, context.chainID, params.receivingChainID, params.messageFee, params.data);
        }
        catch (error) {
            return {
                status: state_machine_1.VerifyStatus.FAIL,
                error: error,
            };
        }
        return {
            status: state_machine_1.VerifyStatus.OK,
        };
    }
    async execute(context) {
        const { params } = context;
        await this._internalMethod.transferCrossChain(context.getMethodContext(), context.transaction.senderAddress, params.recipientAddress, params.nftID, params.receivingChainID, params.messageFee, params.data, params.includeAttributes, context.header.timestamp);
    }
}
exports.TransferCrossChainCommand = TransferCrossChainCommand;
