// lisk-framework/dist-node/modules/dynamic_reward/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExpectedValidatorRewardsResponseSchema = exports.getExpectedValidatorRewardsRequestSchema = exports.configSchema = void 0;
const schemas_1 = require("../reward/schemas");
exports.configSchema = {
    $id: '/dynamicReward/config',
    type: 'object',
    properties: {
        ...schemas_1.configSchema.properties,
        factorMinimumRewardActiveValidators: {
            type: 'integer',
            minimum: 1,
        },
    },
    required: [...schemas_1.configSchema.required, 'factorMinimumRewardActiveValidators'],
};
exports.getExpectedValidatorRewardsRequestSchema = {
    $id: '/dynamicReward/endpoint/getExpectedValidatorRewardsRequestSchema',
    type: 'object',
    required: ['validatorAddress'],
    properties: {
        validatorAddress: {
            type: 'string',
            format: 'lisk32',
        },
    },
};
exports.getExpectedValidatorRewardsResponseSchema = {
    $id: '/dynamicReward/endpoint/getExpectedValidatorRewardsResponseSchema',
    type: 'object',
    required: ['blockReward', 'dailyReward', 'monthlyReward', 'yearlyReward'],
    properties: {
        blockReward: {
            type: 'string',
        },
        dailyReward: {
            type: 'string',
        },
        monthlyReward: {
            type: 'string',
        },
        yearlyReward: {
            type: 'string',
        },
    },
};
