// lisk-framework/dist-node/modules/pos/stores/snapshot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnapshotStore = exports.snapshotStoreSchema = void 0;
const base_store_1 = require("../../base_store");
exports.snapshotStoreSchema = {
    $id: '/pos/store/snapshot',
    type: 'object',
    required: ['validatorWeightSnapshot'],
    properties: {
        validatorWeightSnapshot: {
            type: 'array',
            fieldNumber: 1,
            items: {
                type: 'object',
                required: ['address', 'weight'],
                properties: {
                    address: {
                        dataType: 'bytes',
                        fieldNumber: 1,
                        format: 'lisk32',
                    },
                    weight: {
                        dataType: 'uint64',
                        fieldNumber: 2,
                    },
                },
            },
        },
    },
};
class SnapshotStore extends base_store_1.BaseStore {
    constructor() {
        super(...arguments);
        this.schema = exports.snapshotStoreSchema;
    }
}
exports.SnapshotStore = SnapshotStore;
