// lisk-framework/dist-node/modules/pos/stores/validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorStore = exports.validatorStoreSchema = void 0;
const base_store_1 = require("../../base_store");
const constants_1 = require("../constants");
exports.validatorStoreSchema = {
    $id: '/pos/validator',
    type: 'object',
    required: [
        'name',
        'totalStake',
        'selfStake',
        'lastGeneratedHeight',
        'isBanned',
        'reportMisbehaviorHeights',
        'consecutiveMissedBlocks',
        'commission',
        'lastCommissionIncreaseHeight',
        'sharingCoefficients',
    ],
    properties: {
        name: {
            dataType: 'string',
            fieldNumber: 1,
        },
        totalStake: {
            dataType: 'uint64',
            fieldNumber: 2,
        },
        selfStake: {
            dataType: 'uint64',
            fieldNumber: 3,
        },
        lastGeneratedHeight: {
            dataType: 'uint32',
            fieldNumber: 4,
        },
        isBanned: {
            dataType: 'boolean',
            fieldNumber: 5,
        },
        reportMisbehaviorHeights: {
            type: 'array',
            fieldNumber: 6,
            items: { dataType: 'uint32' },
        },
        consecutiveMissedBlocks: {
            dataType: 'uint32',
            fieldNumber: 7,
        },
        commission: {
            dataType: 'uint32',
            fieldNumber: 8,
        },
        lastCommissionIncreaseHeight: {
            dataType: 'uint32',
            fieldNumber: 9,
        },
        sharingCoefficients: {
            type: 'array',
            fieldNumber: 10,
            items: {
                type: 'object',
                required: ['tokenID', 'coefficient'],
                properties: {
                    tokenID: {
                        dataType: 'bytes',
                        fieldNumber: 1,
                        minLength: constants_1.TOKEN_ID_LENGTH,
                        maxLength: constants_1.TOKEN_ID_LENGTH,
                    },
                    coefficient: {
                        dataType: 'bytes',
                        fieldNumber: 2,
                        maxLength: constants_1.MAX_NUMBER_BYTES_Q96,
                    },
                },
            },
        },
    },
};
class ValidatorStore extends base_store_1.BaseStore {
    constructor() {
        super(...arguments);
        this.schema = exports.validatorStoreSchema;
    }
}
exports.ValidatorStore = ValidatorStore;
