// lisk-framework/dist-node/modules/random/stores/validator_reveals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorRevealsStore = exports.seedRevealSchema = void 0;
const base_store_1 = require("../../base_store");
const constants_1 = require("../../token/constants");
const constants_2 = require("../constants");
exports.seedRevealSchema = {
    $id: '/modules/random/seedReveal',
    type: 'object',
    required: ['validatorReveals'],
    properties: {
        validatorReveals: {
            type: 'array',
            fieldNumber: 1,
            items: {
                type: 'object',
                required: ['generatorAddress', 'seedReveal', 'height', 'valid'],
                properties: {
                    generatorAddress: {
                        dataType: 'bytes',
                        minLength: constants_1.ADDRESS_LENGTH,
                        maxLength: constants_1.ADDRESS_LENGTH,
                        fieldNumber: 1,
                    },
                    seedReveal: {
                        dataType: 'bytes',
                        minLength: constants_2.SEED_LENGTH,
                        maxLength: constants_2.SEED_LENGTH,
                        fieldNumber: 2,
                    },
                    height: {
                        dataType: 'uint32',
                        fieldNumber: 3,
                    },
                    valid: {
                        dataType: 'boolean',
                        fieldNumber: 4,
                    },
                },
            },
        },
    },
};
class ValidatorRevealsStore extends base_store_1.BaseStore {
    constructor() {
        super(...arguments);
        this.schema = exports.seedRevealSchema;
    }
}
exports.ValidatorRevealsStore = ValidatorRevealsStore;
