// @liskhq/lisk-chain/dist-node/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NAME_REGEX = exports.EVENT_ID_LENGTH_BYTES = exports.MAX_EVENTS_PER_BLOCK = exports.EVENT_KEY_LENGTH = exports.EVENT_MAX_TOPICS_PER_EVENT = exports.EVENT_TOTAL_INDEX_LENGTH_BYTES = exports.MAX_PARAMS_SIZE = exports.EVENT_MAX_EVENT_SIZE_BYTES = exports.EVENT_TOPIC_INDEX_LENGTH_BITS = exports.EVENT_INDEX_LENGTH_BITS = exports.EVENT_TOPIC_HASH_LENGTH_BYTES = exports.SMT_PREFIX_SIZE = exports.MAX_CROSS_CHAIN_COMMAND_NAME_LENGTH = exports.MIN_CROSS_CHAIN_COMMAND_NAME_LENGTH = exports.MAX_EVENT_NAME_LENGTH = exports.MIN_EVENT_NAME_LENGTH = exports.MAX_MODULE_NAME_LENGTH = exports.MIN_MODULE_NAME_LENGTH = exports.SIGNATURE_LENGTH_BYTES = exports.MAX_ASSET_DATA_SIZE_BYTES = exports.TAG_TRANSACTION = exports.TAG_BLOCK_HEADER = exports.GENESIS_BLOCK_TRANSACTION_ROOT = exports.GENESIS_BLOCK_SIGNATURE = exports.GENESIS_BLOCK_REWARD = exports.GENESIS_BLOCK_GENERATOR_ADDRESS = exports.GENESIS_BLOCK_VERSION = exports.MAX_UINT32 = exports.LENGTH_ADDRESS = exports.EMPTY_HASH = exports.EMPTY_BUFFER = exports.DEFAULT_MAX_BLOCK_HEADER_CACHE = exports.DEFAULT_MIN_BLOCK_HEADER_CACHE = exports.DEFAULT_KEEP_EVENTS_FOR_HEIGHTS = void 0;
const lisk_cryptography_1 = require("@liskhq/lisk-cryptography");
exports.DEFAULT_KEEP_EVENTS_FOR_HEIGHTS = 300;
exports.DEFAULT_MIN_BLOCK_HEADER_CACHE = 309;
exports.DEFAULT_MAX_BLOCK_HEADER_CACHE = 515;
exports.EMPTY_BUFFER = Buffer.alloc(0);
exports.EMPTY_HASH = lisk_cryptography_1.utils.hash(exports.EMPTY_BUFFER);
exports.LENGTH_ADDRESS = 20;
exports.MAX_UINT32 = 4294967295;
exports.GENESIS_BLOCK_VERSION = 0;
exports.GENESIS_BLOCK_GENERATOR_ADDRESS = exports.EMPTY_BUFFER;
exports.GENESIS_BLOCK_REWARD = BigInt(0);
exports.GENESIS_BLOCK_SIGNATURE = exports.EMPTY_BUFFER;
exports.GENESIS_BLOCK_TRANSACTION_ROOT = exports.EMPTY_HASH;
exports.TAG_BLOCK_HEADER = lisk_cryptography_1.utils.createMessageTag('BH');
exports.TAG_TRANSACTION = lisk_cryptography_1.utils.createMessageTag('TX');
exports.MAX_ASSET_DATA_SIZE_BYTES = 18;
exports.SIGNATURE_LENGTH_BYTES = 64;
exports.MIN_MODULE_NAME_LENGTH = 1;
exports.MAX_MODULE_NAME_LENGTH = 32;
exports.MIN_EVENT_NAME_LENGTH = 1;
exports.MAX_EVENT_NAME_LENGTH = 32;
exports.MIN_CROSS_CHAIN_COMMAND_NAME_LENGTH = 1;
exports.MAX_CROSS_CHAIN_COMMAND_NAME_LENGTH = 32;
exports.SMT_PREFIX_SIZE = 6;
exports.EVENT_TOPIC_HASH_LENGTH_BYTES = 8;
exports.EVENT_INDEX_LENGTH_BITS = 30;
exports.EVENT_TOPIC_INDEX_LENGTH_BITS = 2;
exports.EVENT_MAX_EVENT_SIZE_BYTES = 10 * 1024;
exports.MAX_PARAMS_SIZE = 14 * 1024;
exports.EVENT_TOTAL_INDEX_LENGTH_BYTES = (exports.EVENT_INDEX_LENGTH_BITS + exports.EVENT_TOPIC_INDEX_LENGTH_BITS) / 8;
exports.EVENT_MAX_TOPICS_PER_EVENT = 2 ** exports.EVENT_TOPIC_INDEX_LENGTH_BITS;
exports.EVENT_KEY_LENGTH = exports.EVENT_TOPIC_HASH_LENGTH_BYTES + exports.EVENT_TOTAL_INDEX_LENGTH_BYTES;
exports.MAX_EVENTS_PER_BLOCK = 2 ** exports.EVENT_INDEX_LENGTH_BITS;
exports.EVENT_ID_LENGTH_BYTES = 4 + exports.EVENT_TOTAL_INDEX_LENGTH_BYTES;
exports.NAME_REGEX = /^[a-zA-Z0-9]*$/;
