// lisk-framework/dist-node/modules/auth/events/invalid_signature.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidSignatureEvent = exports.invalidSigDataSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.invalidSigDataSchema = {
    $id: '/auth/events/invalidSigData',
    type: 'object',
    required: [
        'numberOfSignatures',
        'mandatoryKeys',
        'optionalKeys',
        'failingPublicKey',
        'failingSignature',
    ],
    properties: {
        numberOfSignatures: {
            dataType: 'uint32',
            fieldNumber: 1,
        },
        mandatoryKeys: {
            type: 'array',
            items: {
                dataType: 'bytes',
                minLength: constants_1.ED25519_PUBLIC_KEY_LENGTH,
                maxLength: constants_1.ED25519_PUBLIC_KEY_LENGTH,
            },
            fieldNumber: 2,
        },
        optionalKeys: {
            type: 'array',
            items: {
                dataType: 'bytes',
                minLength: constants_1.ED25519_PUBLIC_KEY_LENGTH,
                maxLength: constants_1.ED25519_PUBLIC_KEY_LENGTH,
            },
            fieldNumber: 3,
        },
        failingPublicKey: {
            dataType: 'bytes',
            minLength: constants_1.ED25519_PUBLIC_KEY_LENGTH,
            maxLength: constants_1.ED25519_PUBLIC_KEY_LENGTH,
            fieldNumber: 4,
        },
        failingSignature: {
            dataType: 'bytes',
            minLength: constants_1.ED25519_SIGNATURE_LENGTH,
            maxLength: constants_1.ED25519_SIGNATURE_LENGTH,
            fieldNumber: 5,
        },
    },
};
class InvalidSignatureEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.invalidSigDataSchema;
    }
    error(ctx, senderAddress, data) {
        this.add(ctx, data, [senderAddress], true);
    }
}
exports.InvalidSignatureEvent = InvalidSignatureEvent;
