// lisk-framework/dist-node/modules/pos/events/commission_change.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommissionChangeEvent = exports.commissionChangeEventParams = void 0;
const base_event_1 = require("../../base_event");
exports.commissionChangeEventParams = {
    $id: '/pos/events/commissionChangeData',
    type: 'object',
    required: ['validatorAddress', 'oldCommission', 'newCommission'],
    properties: {
        validatorAddress: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 1,
        },
        oldCommission: {
            dataType: 'uint32',
            fieldNumber: 2,
        },
        newCommission: {
            dataType: 'uint32',
            fieldNumber: 3,
        },
    },
};
class CommissionChangeEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.commissionChangeEventParams;
    }
    log(ctx, data) {
        this.add(ctx, { ...data }, [data.validatorAddress]);
    }
}
exports.CommissionChangeEvent = CommissionChangeEvent;
