// lisk-framework/dist-node/modules/pos/events/rewards_assigned.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RewardsAssignedEvent = exports.rewardsAssignedEventParams = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.rewardsAssignedEventParams = {
    $id: '/pos/events/rewardsAssignedData',
    type: 'object',
    required: ['stakerAddress', 'validatorAddress', 'tokenID', 'amount'],
    properties: {
        stakerAddress: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 1,
        },
        validatorAddress: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 2,
        },
        tokenID: {
            dataType: 'bytes',
            minLength: constants_1.TOKEN_ID_LENGTH,
            maxLength: constants_1.TOKEN_ID_LENGTH,
            fieldNumber: 3,
        },
        amount: {
            dataType: 'uint64',
            fieldNumber: 4,
        },
    },
};
class RewardsAssignedEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.rewardsAssignedEventParams;
    }
    log(ctx, data) {
        this.add(ctx, { ...data }, [data.stakerAddress]);
    }
}
exports.RewardsAssignedEvent = RewardsAssignedEvent;
