// lisk-framework/dist-node/modules/pos/events/validator_staked.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorStakedEvent = exports.validatorStakedDataSchema = void 0;
const base_event_1 = require("../../base_event");
exports.validatorStakedDataSchema = {
    $id: '/pos/events/validatorStakedData',
    type: 'object',
    required: ['senderAddress', 'validatorAddress', 'amount', 'result'],
    properties: {
        senderAddress: {
            dataType: 'bytes',
            fieldNumber: 1,
            format: 'lisk32',
        },
        validatorAddress: {
            dataType: 'bytes',
            fieldNumber: 2,
            format: 'lisk32',
        },
        amount: {
            dataType: 'sint64',
            fieldNumber: 3,
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 4,
        },
    },
};
class ValidatorStakedEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.validatorStakedDataSchema;
    }
    log(ctx, data) {
        this.add(ctx, { ...data, result: 0 }, [
            data.senderAddress,
            data.validatorAddress,
        ]);
    }
    error(ctx, data, result) {
        this.add(ctx, { ...data, result }, [data.senderAddress, data.validatorAddress], true);
    }
}
exports.ValidatorStakedEvent = ValidatorStakedEvent;
