// lisk-framework/dist-node/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MODULE_DB_NAME = exports.STATE_DB_NAME = exports.BFT_BATCH_SIZE = exports.MAX_TRANSACTIONS_SIZE = exports.MAX_NUM_VALIDATORS = exports.MAX_CCM_SIZE = exports.OWNER_READ_WRITE = exports.DEFAULT_MAX_OUTBOUND_CONNECTIONS = exports.DEFAULT_MAX_INBOUND_CONNECTIONS = exports.MAX_BLOCK_CACHE = exports.DEFAULT_PORT_RPC = exports.DEFAULT_PORT_P2P = exports.DEFAULT_HOST = exports.RPC_MODES = exports.APP_EVENT_SHUTDOWN = exports.APP_EVENT_READY = exports.APP_IDENTIFIER = exports.controlReg = exports.actionWithModuleNameReg = exports.moduleNameReg = exports.eventWithModuleNameReg = void 0;
exports.eventWithModuleNameReg = /^[a-zA-Z]\w*_[a-zA-Z][a-zA-Z0-9]*$/;
exports.moduleNameReg = /^[a-zA-Z][a-zA-Z0-9_]*$/;
exports.actionWithModuleNameReg = /^[a-zA-Z][a-zA-Z0-9_]*_[a-zA-Z][a-zA-Z0-9]*$/;
exports.controlReg = /[\x00-\x1F\x7F]/g;
exports.APP_IDENTIFIER = 'app';
exports.APP_EVENT_READY = 'app_ready';
exports.APP_EVENT_SHUTDOWN = 'app_shutdown';
exports.RPC_MODES = {
    IPC: 'ipc',
    WS: 'ws',
    HTTP: 'http',
};
exports.DEFAULT_HOST = '127.0.0.1';
exports.DEFAULT_PORT_P2P = 7667;
exports.DEFAULT_PORT_RPC = 7887;
exports.MAX_BLOCK_CACHE = 515;
exports.DEFAULT_MAX_INBOUND_CONNECTIONS = 100;
exports.DEFAULT_MAX_OUTBOUND_CONNECTIONS = 20;
exports.OWNER_READ_WRITE = 0o600;
exports.MAX_CCM_SIZE = 10240;
exports.MAX_NUM_VALIDATORS = 199;
exports.MAX_TRANSACTIONS_SIZE = 15 * 1024;
exports.BFT_BATCH_SIZE = 103;
exports.STATE_DB_NAME = 'state.db';
exports.MODULE_DB_NAME = 'module.db';
