// lisk-framework/dist-node/system_dirs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.systemDirs = void 0;
const os_1 = require("os");
const path_1 = require("path");
const systemDirs = (dataPath) => {
    const dataPathWithoutTilde = dataPath.replace('~', (0, os_1.homedir)());
    return {
        dataPath: (0, path_1.resolve)(dataPathWithoutTilde),
        data: (0, path_1.resolve)((0, path_1.join)(dataPathWithoutTilde, 'data')),
        config: (0, path_1.resolve)((0, path_1.join)(dataPathWithoutTilde, 'config')),
        tmp: (0, path_1.resolve)((0, path_1.join)(dataPathWithoutTilde, 'tmp')),
        logs: (0, path_1.resolve)((0, path_1.join)(dataPathWithoutTilde, 'logs')),
        sockets: (0, path_1.resolve)((0, path_1.join)(dataPathWithoutTilde, 'tmp', 'sockets')),
        pids: (0, path_1.resolve)((0, path_1.join)(dataPathWithoutTilde, 'tmp', 'pids')),
        plugins: (0, path_1.resolve)((0, path_1.join)(dataPathWithoutTilde, 'plugins')),
    };
};
exports.systemDirs = systemDirs;
