// lisk-framework/dist-node/modules/interoperability/events/ccm_processed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CcmProcessedEvent = exports.ccmProcessedEventSchema = void 0;
const base_event_1 = require("../../base_event");
const schemas_1 = require("../schemas");
exports.ccmProcessedEventSchema = {
    $id: '/interoperability/events/ccmProcessed',
    type: 'object',
    required: ['ccm', 'result', 'code'],
    properties: {
        ccm: {
            fieldNumber: 1,
            type: schemas_1.ccmSchema.type,
            required: [...schemas_1.ccmSchema.required],
            properties: {
                ...schemas_1.ccmSchema.properties,
            },
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 2,
        },
        code: {
            dataType: 'uint32',
            fieldNumber: 3,
        },
    },
};
class CcmProcessedEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.ccmProcessedEventSchema;
    }
    log(ctx, sendingChainID, receivingChainID, data) {
        this.add(ctx, data, [sendingChainID, receivingChainID]);
    }
}
exports.CcmProcessedEvent = CcmProcessedEvent;
