// lisk-framework/dist-node/modules/interoperability/events/ccm_send_success.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CcmSendSuccessEvent = exports.ccmSendSuccessDataSchema = void 0;
const base_event_1 = require("../../base_event");
const schemas_1 = require("../schemas");
exports.ccmSendSuccessDataSchema = {
    $id: '/interoperability/events/ccmSendSuccess',
    type: 'object',
    required: ['ccm'],
    properties: {
        ccm: {
            fieldNumber: 1,
            type: schemas_1.ccmSchema.type,
            required: [...schemas_1.ccmSchema.required],
            properties: {
                ...schemas_1.ccmSchema.properties,
            },
        },
    },
};
class CcmSendSuccessEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.ccmSendSuccessDataSchema;
    }
    log(ctx, sendingChainID, receivingChainID, sentCCMID, data, noRevert = false) {
        this.add(ctx, data, [sendingChainID, receivingChainID, sentCCMID], noRevert);
    }
}
exports.CcmSendSuccessEvent = CcmSendSuccessEvent;
