// lisk-framework/dist-node/modules/interoperability/events/invalid_outbox_root_verification.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidOutboxRootVerificationEvent = exports.invalidOutboxRootVerificationSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.invalidOutboxRootVerificationSchema = {
    $id: '/interoperability/events/invalidOutboxRootVerification',
    type: 'object',
    required: ['inboxRoot', 'partnerChainOutboxRoot'],
    properties: {
        inboxRoot: {
            dataType: 'bytes',
            fieldNumber: 1,
            minLength: constants_1.HASH_LENGTH,
            maxLength: constants_1.HASH_LENGTH,
        },
        partnerChainOutboxRoot: {
            dataType: 'bytes',
            fieldNumber: 2,
            minLength: constants_1.HASH_LENGTH,
            maxLength: constants_1.HASH_LENGTH,
        },
    },
};
class InvalidOutboxRootVerificationEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.invalidOutboxRootVerificationSchema;
    }
    error(ctx, chainID, data) {
        this.add(ctx, data, [chainID], true);
    }
}
exports.InvalidOutboxRootVerificationEvent = InvalidOutboxRootVerificationEvent;
