// lisk-framework/dist-node/modules/interoperability/stores/chain_validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainValidatorsStore = exports.chainValidatorsSchema = void 0;
const base_store_1 = require("../../base_store");
const constants_1 = require("../constants");
exports.chainValidatorsSchema = {
    $id: '/modules/interoperability/chainValidators',
    type: 'object',
    required: ['activeValidators', 'certificateThreshold'],
    properties: {
        activeValidators: {
            type: 'array',
            fieldNumber: 1,
            minItems: 1,
            maxItems: constants_1.MAX_NUM_VALIDATORS,
            items: {
                type: 'object',
                required: ['blsKey', 'bftWeight'],
                properties: {
                    blsKey: {
                        dataType: 'bytes',
                        minLength: constants_1.BLS_PUBLIC_KEY_LENGTH,
                        maxLength: constants_1.BLS_PUBLIC_KEY_LENGTH,
                        fieldNumber: 1,
                    },
                    bftWeight: {
                        dataType: 'uint64',
                        fieldNumber: 2,
                    },
                },
            },
        },
        certificateThreshold: {
            dataType: 'uint64',
            fieldNumber: 2,
        },
    },
};
class ChainValidatorsStore extends base_store_1.BaseStore {
    constructor() {
        super(...arguments);
        this.schema = exports.chainValidatorsSchema;
    }
    get storePrefix() {
        return constants_1.STORE_PREFIX;
    }
}
exports.ChainValidatorsStore = ChainValidatorsStore;
