// lisk-framework/dist-node/modules/interoperability/stores/own_chain_account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OwnChainAccountStore = exports.ownChainAccountSchema = void 0;
const base_store_1 = require("../../base_store");
const constants_1 = require("../../token/constants");
const constants_2 = require("../constants");
exports.ownChainAccountSchema = {
    $id: '/modules/interoperability/ownChainAccount',
    type: 'object',
    required: ['name', 'chainID', 'nonce'],
    properties: {
        name: {
            dataType: 'string',
            minLength: constants_2.MIN_CHAIN_NAME_LENGTH,
            maxLength: constants_2.MAX_CHAIN_NAME_LENGTH,
            fieldNumber: 1,
        },
        chainID: {
            dataType: 'bytes',
            minLength: constants_1.CHAIN_ID_LENGTH,
            maxLength: constants_1.CHAIN_ID_LENGTH,
            fieldNumber: 2,
        },
        nonce: {
            dataType: 'uint64',
            fieldNumber: 3,
        },
    },
};
class OwnChainAccountStore extends base_store_1.BaseStore {
    constructor() {
        super(...arguments);
        this.schema = exports.ownChainAccountSchema;
    }
    get storePrefix() {
        return constants_2.STORE_PREFIX;
    }
}
exports.OwnChainAccountStore = OwnChainAccountStore;
