// lisk-framework/dist-node/modules/interoperability/stores/terminated_outbox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerminatedOutboxStore = exports.terminatedOutboxSchema = void 0;
const base_store_1 = require("../../base_store");
const constants_1 = require("../constants");
exports.terminatedOutboxSchema = {
    $id: '/modules/interoperability/terminatedOutbox',
    type: 'object',
    required: ['outboxRoot', 'outboxSize', 'partnerChainInboxSize'],
    properties: {
        outboxRoot: {
            dataType: 'bytes',
            minLength: constants_1.HASH_LENGTH,
            maxLength: constants_1.HASH_LENGTH,
            fieldNumber: 1,
        },
        outboxSize: {
            dataType: 'uint32',
            fieldNumber: 2,
        },
        partnerChainInboxSize: {
            dataType: 'uint32',
            fieldNumber: 3,
        },
    },
};
class TerminatedOutboxStore extends base_store_1.BaseStore {
    constructor() {
        super(...arguments);
        this.schema = exports.terminatedOutboxSchema;
    }
    get storePrefix() {
        return constants_1.STORE_PREFIX;
    }
}
exports.TerminatedOutboxStore = TerminatedOutboxStore;
