// lisk-framework/dist-node/modules/interoperability/stores/terminated_state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerminatedStateStore = exports.terminatedStateSchema = void 0;
const base_store_1 = require("../../base_store");
const constants_1 = require("../constants");
exports.terminatedStateSchema = {
    $id: '/modules/interoperability/terminatedState',
    type: 'object',
    required: ['stateRoot', 'mainchainStateRoot', 'initialized'],
    properties: {
        stateRoot: {
            dataType: 'bytes',
            minLength: constants_1.HASH_LENGTH,
            maxLength: constants_1.HASH_LENGTH,
            fieldNumber: 1,
        },
        mainchainStateRoot: {
            dataType: 'bytes',
            minLength: constants_1.HASH_LENGTH,
            maxLength: constants_1.HASH_LENGTH,
            fieldNumber: 2,
        },
        initialized: {
            dataType: 'boolean',
            fieldNumber: 3,
        },
    },
};
class TerminatedStateStore extends base_store_1.BaseStore {
    constructor() {
        super(...arguments);
        this.schema = exports.terminatedStateSchema;
    }
    get storePrefix() {
        return constants_1.STORE_PREFIX;
    }
}
exports.TerminatedStateStore = TerminatedStateStore;
