// lisk-framework/dist-node/engine/consensus/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EMPTY_HASH = exports.NETWORK_LEGACY_GET_BLOCKS_FROM_ID = exports.NETWORK_RPC_GET_HIGHEST_COMMON_BLOCK = exports.NETWORK_RPC_GET_BLOCKS_FROM_ID = exports.NETWORK_RPC_GET_LAST_BLOCK = exports.NETWORK_EVENT_POST_NODE_INFO = exports.NETWORK_EVENT_POST_BLOCK = exports.CONSENSUS_EVENT_VALIDATORS_CHANGED = exports.CONSENSUS_EVENT_FINALIZED_HEIGHT_CHANGED = exports.CONSENSUS_EVENT_BLOCK_DELETE = exports.CONSENSUS_EVENT_NETWORK_BLOCK_NEW = exports.CONSENSUS_EVENT_BLOCK_NEW = exports.CONSENSUS_EVENT_BLOCK_BROADCAST = exports.CONSENSUS_EVENT_FORK_DETECTED = void 0;
const lisk_cryptography_1 = require("@liskhq/lisk-cryptography");
exports.CONSENSUS_EVENT_FORK_DETECTED = 'CONSENSUS_EVENT_FORK_DETECTED';
exports.CONSENSUS_EVENT_BLOCK_BROADCAST = 'CONSENSUS_EVENT_BLOCK_BROADCAST';
exports.CONSENSUS_EVENT_BLOCK_NEW = 'CONSENSUS_EVENT_BLOCK_NEW';
exports.CONSENSUS_EVENT_NETWORK_BLOCK_NEW = 'CONSENSUS_EVENT_NETWORK_BLOCK_NEW';
exports.CONSENSUS_EVENT_BLOCK_DELETE = 'CONSENSUS_EVENT_BLOCK_DELETE';
exports.CONSENSUS_EVENT_FINALIZED_HEIGHT_CHANGED = 'CONSENSUS_EVENT_FINALIZED_HEIGHT_CHANGED';
exports.CONSENSUS_EVENT_VALIDATORS_CHANGED = 'CONSENSUS_EVENT_VALIDATORS_CHANGED';
exports.NETWORK_EVENT_POST_BLOCK = 'postBlock';
exports.NETWORK_EVENT_POST_NODE_INFO = 'postNodeInfo';
exports.NETWORK_RPC_GET_LAST_BLOCK = 'getLastBlock';
exports.NETWORK_RPC_GET_BLOCKS_FROM_ID = 'getBlocksFromId';
exports.NETWORK_RPC_GET_HIGHEST_COMMON_BLOCK = 'getHighestCommonBlock';
exports.NETWORK_LEGACY_GET_BLOCKS_FROM_ID = 'getLegacyBlocksFromId';
exports.EMPTY_HASH = lisk_cryptography_1.utils.hash(Buffer.alloc(0));
