// lisk-framework/dist-node/modules/fee/events/generator_fee_processed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeneratorFeeProcessedEvent = exports.generatorFeeProcessedSchema = void 0;
const base_event_1 = require("../../base_event");
exports.generatorFeeProcessedSchema = {
    $id: '/fee/events/generatorFeeProcessed',
    type: 'object',
    required: ['senderAddress', 'generatorAddress', 'burntAmount', 'generatorAmount'],
    properties: {
        senderAddress: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 1,
        },
        generatorAddress: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 2,
        },
        burntAmount: {
            dataType: 'uint64',
            fieldNumber: 3,
        },
        generatorAmount: {
            dataType: 'uint64',
            fieldNumber: 4,
        },
    },
};
class GeneratorFeeProcessedEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.generatorFeeProcessedSchema;
    }
    log(ctx, data) {
        this.add(ctx, data, [data.senderAddress, data.generatorAddress]);
    }
}
exports.GeneratorFeeProcessedEvent = GeneratorFeeProcessedEvent;
