// lisk-framework/dist-node/modules/fee/events/relayer_fee_processed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelayerFeeProcessedEvent = exports.relayerFeeProcessedSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.relayerFeeProcessedSchema = {
    $id: '/fee/events/relayerFeeProcessed',
    type: 'object',
    required: ['ccmID', 'relayerAddress', 'burntAmount', 'relayerAmount'],
    properties: {
        ccmID: {
            dataType: 'bytes',
            minLength: constants_1.HASH_LENGTH,
            maxLength: constants_1.HASH_LENGTH,
            fieldNumber: 1,
        },
        relayerAddress: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 2,
        },
        burntAmount: {
            dataType: 'uint64',
            fieldNumber: 3,
        },
        relayerAmount: {
            dataType: 'uint64',
            fieldNumber: 4,
        },
    },
};
class RelayerFeeProcessedEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.relayerFeeProcessedSchema;
    }
    log(ctx, data) {
        this.add(ctx, data, [data.relayerAddress]);
    }
}
exports.RelayerFeeProcessedEvent = RelayerFeeProcessedEvent;
