// lisk-framework/dist-node/engine/generator/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GENERATOR_EVENT_NEW_TRANSACTION_ANNOUNCEMENT = exports.GENERATOR_EVENT_NEW_TRANSACTION = exports.GENESIS_BLOCK_VERSION = exports.EMPTY_HASH = exports.EMPTY_BUFFER = exports.GENERATOR_STORE_KEY_PREFIX = exports.GENERATOR_STORE_INFO_PREFIX = exports.NETWORK_EVENT_POST_TRANSACTIONS_ANNOUNCEMENT = exports.NETWORK_RPC_GET_TRANSACTIONS = exports.LOAD_TRANSACTION_RETRIES = exports.FORGE_INTERVAL = exports.DEFAULT_RATE_LIMIT_FREQUENCY = exports.DEFAULT_RELEASE_INTERVAL = exports.DEFAULT_RELEASE_LIMIT = void 0;
const lisk_cryptography_1 = require("@liskhq/lisk-cryptography");
exports.DEFAULT_RELEASE_LIMIT = 100;
exports.DEFAULT_RELEASE_INTERVAL = 5000;
exports.DEFAULT_RATE_LIMIT_FREQUENCY = 3;
exports.FORGE_INTERVAL = 1000;
exports.LOAD_TRANSACTION_RETRIES = 5;
exports.NETWORK_RPC_GET_TRANSACTIONS = 'getTransactions';
exports.NETWORK_EVENT_POST_TRANSACTIONS_ANNOUNCEMENT = 'postTransactionsAnnouncement';
exports.GENERATOR_STORE_INFO_PREFIX = lisk_cryptography_1.utils.intToBuffer(0, 4);
exports.GENERATOR_STORE_KEY_PREFIX = lisk_cryptography_1.utils.intToBuffer(1, 4);
exports.EMPTY_BUFFER = Buffer.alloc(0);
exports.EMPTY_HASH = lisk_cryptography_1.utils.hash(Buffer.alloc(0));
exports.GENESIS_BLOCK_VERSION = 0;
exports.GENERATOR_EVENT_NEW_TRANSACTION = 'GENERATOR_EVENT_NEW_TRANSACTION';
exports.GENERATOR_EVENT_NEW_TRANSACTION_ANNOUNCEMENT = 'GENERATOR_EVENT_NEW_TRANSACTION_ANNOUNCEMENT';
