// lisk-framework/dist-node/modules/validators/stores/validators_params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorsParamsStore = exports.validatorsParamsSchema = exports.validatorSchema = void 0;
const base_store_1 = require("../../base_store");
exports.validatorSchema = {
    type: 'object',
    required: ['address', 'bftWeight', 'generatorKey', 'blsKey'],
    properties: {
        address: {
            fieldNumber: 1,
            dataType: 'bytes',
            format: 'lisk32',
        },
        bftWeight: {
            fieldNumber: 2,
            dataType: 'uint64',
        },
        generatorKey: {
            fieldNumber: 3,
            dataType: 'bytes',
        },
        blsKey: {
            fieldNumber: 4,
            dataType: 'bytes',
        },
    },
};
exports.validatorsParamsSchema = {
    $id: '/validators/validatorsParams',
    type: 'object',
    required: ['validators', 'preCommitThreshold', 'certificateThreshold'],
    properties: {
        preCommitThreshold: {
            fieldNumber: 1,
            dataType: 'uint64',
        },
        certificateThreshold: {
            fieldNumber: 2,
            dataType: 'uint64',
        },
        validators: {
            fieldNumber: 3,
            type: 'array',
            items: {
                ...exports.validatorSchema,
            },
        },
    },
};
class ValidatorsParamsStore extends base_store_1.BaseStore {
    constructor() {
        super(...arguments);
        this.schema = exports.validatorsParamsSchema;
    }
}
exports.ValidatorsParamsStore = ValidatorsParamsStore;
