// lisk-framework/dist-node/modules/validators/stores/validator_keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorKeysStore = exports.validatorKeysSchema = void 0;
const base_store_1 = require("../../base_store");
const constants_1 = require("../constants");
exports.validatorKeysSchema = {
    $id: '/validators/validatorAccountSubStore',
    title: 'Validators Account Keys',
    type: 'object',
    properties: {
        generatorKey: {
            dataType: 'bytes',
            fieldNumber: 1,
            minLength: constants_1.ED25519_PUBLIC_KEY_LENGTH,
            maxLength: constants_1.ED25519_PUBLIC_KEY_LENGTH,
        },
        blsKey: {
            dataType: 'bytes',
            fieldNumber: 2,
            minLength: constants_1.BLS_PUBLIC_KEY_LENGTH,
            maxLength: constants_1.BLS_PUBLIC_KEY_LENGTH,
        },
    },
    required: ['generatorKey', 'blsKey'],
};
class ValidatorKeysStore extends base_store_1.BaseStore {
    constructor() {
        super(...arguments);
        this.schema = exports.validatorKeysSchema;
    }
}
exports.ValidatorKeysStore = ValidatorKeysStore;
