// lisk-framework/dist-node/modules/token/events/all_tokens_from_chain_supported_removed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllTokensFromChainSupportRemovedEvent = exports.allTokensFromChainSupportRemovedEventSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.allTokensFromChainSupportRemovedEventSchema = {
    $id: '/token/events/allTokensFromChainSupportRemoved',
    type: 'object',
    required: ['chainID'],
    properties: {
        chainID: {
            dataType: 'bytes',
            minLength: constants_1.CHAIN_ID_LENGTH,
            maxLength: constants_1.CHAIN_ID_LENGTH,
            fieldNumber: 1,
        },
    },
};
class AllTokensFromChainSupportRemovedEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.allTokensFromChainSupportRemovedEventSchema;
    }
    log(ctx, chainID) {
        this.add(ctx, { chainID }, [chainID]);
    }
}
exports.AllTokensFromChainSupportRemovedEvent = AllTokensFromChainSupportRemovedEvent;
