// lisk-framework/dist-node/modules/token/events/before_ccm_forwarding.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BeforeCCMForwardingEvent = exports.beforeCCMForwardingEventSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.beforeCCMForwardingEventSchema = {
    $id: '/token/events/beforeCCMForwarding',
    type: 'object',
    required: ['ccmID', 'messageFeeTokenID', 'result'],
    properties: {
        ccmID: {
            dataType: 'bytes',
            minLength: constants_1.HASH_LENGTH,
            maxLength: constants_1.HASH_LENGTH,
            fieldNumber: 1,
        },
        messageFeeTokenID: {
            dataType: 'bytes',
            minLength: constants_1.TOKEN_ID_LENGTH,
            maxLength: constants_1.TOKEN_ID_LENGTH,
            fieldNumber: 2,
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 3,
        },
    },
};
class BeforeCCMForwardingEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.beforeCCMForwardingEventSchema;
    }
    log(ctx, sendingChainID, receivingChainID, data) {
        this.add(ctx, { ...data, result: 0 }, [
            sendingChainID,
            receivingChainID,
        ]);
    }
    error(ctx, sendingChainID, receivingChainID, data, result) {
        this.add(ctx, { ...data, result }, [sendingChainID, receivingChainID], true);
    }
}
exports.BeforeCCMForwardingEvent = BeforeCCMForwardingEvent;
