// lisk-framework/dist-node/modules/token/events/initialize_escrow_account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InitializeEscrowAccountEvent = exports.initializeEscrowAccountEventSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.initializeEscrowAccountEventSchema = {
    $id: '/token/events/initializeEscrowAccount',
    type: 'object',
    required: ['chainID', 'tokenID', 'initializationFee', 'result'],
    properties: {
        chainID: {
            dataType: 'bytes',
            minLength: constants_1.CHAIN_ID_LENGTH,
            maxLength: constants_1.CHAIN_ID_LENGTH,
            fieldNumber: 1,
        },
        tokenID: {
            dataType: 'bytes',
            minLength: constants_1.TOKEN_ID_LENGTH,
            maxLength: constants_1.TOKEN_ID_LENGTH,
            fieldNumber: 2,
        },
        initializationFee: {
            dataType: 'uint64',
            fieldNumber: 3,
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 4,
        },
    },
};
class InitializeEscrowAccountEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.initializeEscrowAccountEventSchema;
    }
    log(ctx, data) {
        this.add(ctx, { ...data, result: 0 }, [data.chainID]);
    }
    error(ctx, data, result) {
        this.add(ctx, { ...data, result }, [data.chainID], true);
    }
}
exports.InitializeEscrowAccountEvent = InitializeEscrowAccountEvent;
