// lisk-framework/dist-node/modules/token/events/lock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LockEvent = exports.lockEventSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.lockEventSchema = {
    $id: '/token/events/lock',
    type: 'object',
    required: ['address', 'module', 'tokenID', 'amount', 'result'],
    properties: {
        address: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 1,
        },
        module: {
            dataType: 'string',
            minLength: constants_1.MIN_MODULE_NAME_LENGTH,
            maxLength: constants_1.MAX_MODULE_NAME_LENGTH,
            fieldNumber: 2,
        },
        tokenID: {
            dataType: 'bytes',
            minLength: constants_1.TOKEN_ID_LENGTH,
            maxLength: constants_1.TOKEN_ID_LENGTH,
            fieldNumber: 3,
        },
        amount: {
            dataType: 'uint64',
            fieldNumber: 4,
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 5,
        },
    },
};
class LockEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.lockEventSchema;
    }
    log(ctx, data) {
        this.add(ctx, { ...data, result: 0 }, [data.address]);
    }
    error(ctx, data, result) {
        this.add(ctx, { ...data, result }, [data.address], true);
    }
}
exports.LockEvent = LockEvent;
