// lisk-framework/dist-node/modules/token/events/transfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferEvent = exports.transferEventSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.transferEventSchema = {
    $id: '/token/events/transfer',
    type: 'object',
    required: ['senderAddress', 'recipientAddress', 'tokenID', 'amount', 'result'],
    properties: {
        senderAddress: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 1,
        },
        recipientAddress: {
            dataType: 'bytes',
            format: 'lisk32',
            fieldNumber: 2,
        },
        tokenID: {
            dataType: 'bytes',
            minLength: constants_1.TOKEN_ID_LENGTH,
            maxLength: constants_1.TOKEN_ID_LENGTH,
            fieldNumber: 3,
        },
        amount: {
            dataType: 'uint64',
            fieldNumber: 4,
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 5,
        },
    },
};
class TransferEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.transferEventSchema;
    }
    log(ctx, data) {
        this.add(ctx, { ...data, result: 0 }, [
            data.senderAddress,
            data.recipientAddress,
        ]);
    }
    error(ctx, data, result) {
        this.add(ctx, { ...data, result }, [data.senderAddress, data.recipientAddress], true);
    }
}
exports.TransferEvent = TransferEvent;
