// lisk-framework/dist-node/modules/validators/events/generator_key_registration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeneratorKeyRegistrationEvent = exports.generatorKeyRegDataSchema = void 0;
const base_event_1 = require("../../base_event");
const constants_1 = require("../constants");
exports.generatorKeyRegDataSchema = {
    $id: '/validators/event/generatorKeyRegData',
    type: 'object',
    required: ['generatorKey', 'result'],
    properties: {
        generatorKey: {
            dataType: 'bytes',
            minLength: constants_1.ED25519_PUBLIC_KEY_LENGTH,
            maxLength: constants_1.ED25519_PUBLIC_KEY_LENGTH,
            fieldNumber: 1,
        },
        result: {
            dataType: 'uint32',
            fieldNumber: 2,
        },
    },
};
class GeneratorKeyRegistrationEvent extends base_event_1.BaseEvent {
    constructor() {
        super(...arguments);
        this.schema = exports.generatorKeyRegDataSchema;
    }
    log(ctx, validatorAddress, data) {
        const noRevert = data.result !== 0;
        this.add(ctx, data, [validatorAddress], noRevert);
    }
}
exports.GeneratorKeyRegistrationEvent = GeneratorKeyRegistrationEvent;
