// lisk-framework/dist-node/modules/poa/stores/snapshot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnapshotStore = exports.snapshotSchema = void 0;
const base_store_1 = require("../../base_store");
const constants_1 = require("../constants");
exports.snapshotSchema = {
    $id: '/poa/snapshot',
    type: 'object',
    required: ['validators', 'threshold'],
    properties: {
        validators: {
            type: 'array',
            fieldNumber: 1,
            items: {
                type: 'object',
                required: ['address', 'weight'],
                properties: {
                    address: {
                        dataType: 'bytes',
                        minLength: constants_1.NUM_BYTES_ADDRESS,
                        maxLength: constants_1.NUM_BYTES_ADDRESS,
                        fieldNumber: 1,
                    },
                    weight: {
                        dataType: 'uint64',
                        fieldNumber: 2,
                    },
                },
            },
        },
        threshold: {
            dataType: 'uint64',
            fieldNumber: 2,
        },
    },
};
class SnapshotStore extends base_store_1.BaseStore {
    constructor() {
        super(...arguments);
        this.schema = exports.snapshotSchema;
    }
}
exports.SnapshotStore = SnapshotStore;
