// lisk-framework/dist-node/modules/reward/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONTEXT_STORE_KEY_BLOCK_REDUCTION = exports.CONTEXT_STORE_KEY_BLOCK_REWARD = exports.defaultConfig = exports.REWARD_REDUCTION_FACTOR_BFT = exports.REWARD_REDUCTION_NO_ACCOUNT = exports.REWARD_REDUCTION_MAX_PREVOTES = exports.REWARD_REDUCTION_SEED_REVEAL = exports.REWARD_NO_REDUCTION = exports.MODULE_ID_REWARD = void 0;
exports.MODULE_ID_REWARD = 10;
exports.REWARD_NO_REDUCTION = 0;
exports.REWARD_REDUCTION_SEED_REVEAL = 1;
exports.REWARD_REDUCTION_MAX_PREVOTES = 2;
exports.REWARD_REDUCTION_NO_ACCOUNT = 3;
exports.REWARD_REDUCTION_FACTOR_BFT = BigInt(4);
exports.defaultConfig = {
    offset: 2160,
    distance: 3000000,
    brackets: ['500000000', '400000000', '300000000', '200000000', '100000000'],
    rewardReductionFactorBFT: exports.REWARD_REDUCTION_FACTOR_BFT.toString(),
};
exports.CONTEXT_STORE_KEY_BLOCK_REWARD = 'CONTEXT_STORE_KEY_BLOCK_REWARD';
exports.CONTEXT_STORE_KEY_BLOCK_REDUCTION = 'CONTEXT_STORE_KEY_BLOCK_REDUCTION';
