// lisk-framework/dist-node/modules/reward/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRewardTokenIDResponseSchema = exports.getAnnualInflationResponseSchema = exports.getDefaultRewardAtHeightResponseSchema = exports.getAnnualInflationRequestSchema = exports.getDefaultRewardAtHeightRequestSchema = exports.heightSchema = exports.configSchema = void 0;
exports.configSchema = {
    $id: '/reward/config',
    type: 'object',
    properties: {
        tokenID: {
            type: 'string',
            format: 'hex',
            minLength: 16,
            maxLength: 16,
        },
        offset: {
            type: 'integer',
            minimum: 1,
        },
        distance: {
            type: 'integer',
            minimum: 1,
        },
        brackets: {
            type: 'array',
            items: {
                type: 'string',
                format: 'uint64',
            },
        },
    },
    required: ['tokenID', 'offset', 'distance', 'brackets'],
};
exports.heightSchema = {
    $id: '/reward/endpoint/height',
    type: 'object',
    required: ['height'],
    properties: {
        height: {
            type: 'integer',
            format: 'uint32',
        },
    },
};
exports.getDefaultRewardAtHeightRequestSchema = exports.heightSchema;
exports.getAnnualInflationRequestSchema = exports.heightSchema;
exports.getDefaultRewardAtHeightResponseSchema = {
    $id: '/reward/endpoint/getDefaultRewardAtHeightResponse',
    type: 'object',
    required: ['reward'],
    properties: {
        reward: {
            type: 'string',
            format: 'uint64',
        },
    },
};
exports.getAnnualInflationResponseSchema = {
    $id: '/reward/endpoint/getAnnualInflationResponse',
    type: 'object',
    required: ['tokenID', 'rate'],
    properties: {
        tokenID: {
            type: 'string',
            format: 'hex',
        },
        rate: {
            type: 'string',
            format: 'uint64',
            minLength: 16,
            maxLength: 16,
        },
    },
};
exports.getRewardTokenIDResponseSchema = {
    $id: '/reward/endpoint/getRewardTokenID',
    type: 'object',
    required: ['tokenID'],
    properties: {
        tokenID: {
            type: 'string',
            format: 'hex',
            minLength: 16,
            maxLength: 16,
        },
    },
};
