// lisk-framework/dist-node/modules/nft/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAX_LENGTH_DATA = exports.LENGTH_TOKEN_ID = exports.FEE_CREATE_NFT = exports.ALL_SUPPORTED_NFTS_KEY = exports.EMPTY_BYTES = exports.CCM_STATUS_CODE_OK = exports.CROSS_CHAIN_COMMAND_NAME_TRANSFER = exports.NFT_NOT_LOCKED = exports.MODULE_NAME_NFT = exports.LENGTH_ADDRESS = exports.MAX_LENGTH_MODULE_NAME = exports.MIN_LENGTH_MODULE_NAME = exports.LENGTH_INDEX = exports.LENGTH_COLLECTION_ID = exports.LENGTH_NFT_ID = exports.LENGTH_CHAIN_ID = void 0;
exports.LENGTH_CHAIN_ID = 4;
exports.LENGTH_NFT_ID = 16;
exports.LENGTH_COLLECTION_ID = 4;
exports.LENGTH_INDEX = exports.LENGTH_NFT_ID - exports.LENGTH_CHAIN_ID - exports.LENGTH_COLLECTION_ID;
exports.MIN_LENGTH_MODULE_NAME = 1;
exports.MAX_LENGTH_MODULE_NAME = 32;
exports.LENGTH_ADDRESS = 20;
exports.MODULE_NAME_NFT = 'nft';
exports.NFT_NOT_LOCKED = exports.MODULE_NAME_NFT;
exports.CROSS_CHAIN_COMMAND_NAME_TRANSFER = 'crossChainTransfer';
exports.CCM_STATUS_CODE_OK = 0;
exports.EMPTY_BYTES = Buffer.alloc(0);
exports.ALL_SUPPORTED_NFTS_KEY = exports.EMPTY_BYTES;
exports.FEE_CREATE_NFT = 5000000;
exports.LENGTH_TOKEN_ID = 8;
exports.MAX_LENGTH_DATA = 64;
